File: common/src/test/java/dev/number6/db/DynamoDatabaseAdaptorIntegrationTest.java
package dev.number6.db;

import dev.number6.comprehend.results.PresentableEntityResults;
import dev.number6.comprehend.results.PresentableKeyPhrasesResults;
import dev.number6.comprehend.results.PresentableSentimentResults;
import dev.number6.db.model.ChannelComprehensionSummary;
import dev.number6.db.port.DatabasePort;
import dev.number6.generate.CommonRDG;
import org.junit.jupiter.api.Test;

import java.time.LocalDate;
import java.util.Collection;
import java.util.function.Consumer;

import static org.assertj.core.api.Assertions.assertThat;

public class DynamoDatabaseAdaptorIntegrationTest {

    private TestDynamoDatabaseAdaptorComponent component = DaggerTestDynamoDatabaseAdaptorComponent.create();

    @Test
    void createChannels() {
        Collection<String> channelNames = CommonRDG.list(CommonRDG.string()).next();
        TestDynamoDatabaseAdaptorComponent.FakeDynamoDBMapper mapper = (TestDynamoDatabaseAdaptorComponent.FakeDynamoDBMapper) component.getFakeDynamoDBMapper();

        component.getDatabase().createNewSummaryForChannels(channelNames, LocalDate.now());
        assertThat(mapper.getSaved()).hasSize(channelNames.size())
                .extracting("channelName")
                .containsExactlyInAnyOrder(channelNames.toArray());
    }

    @Test
    void saveEntityResults() {
        PresentableEntityResults results = CommonRDG.presentableEntityResults().next();

        ChannelComprehensionSummary summary = saveResults(m -> m.save(results));
        assertThat(summary.getEntityTotals()).isEqualTo(results.getResults());
    }

    @Test
    void saveSentimentResults() {
        PresentableSentimentResults results = CommonRDG.presentableSentimentResults().next();

        ChannelComprehensionSummary summary = saveResults(m -> m.save(results));
        assertThat(summary.getSentimentTotals()).isEqualTo(results.getSentimentTotals());
        assertThat(summary.getSentimentScoreTotals()).isEqualTo(results.getSentimentScoreTotals());
    }

    @Test
    void saveKeyPhrasesResults() {
        PresentableKeyPhrasesResults results = CommonRDG.presentableKeyPhrasesResults().next();

        ChannelComprehensionSummary summary = saveResults(m -> m.save(results));
        assertThat(summary.getKeyPhrasesTotals()).isEqualTo(results.getResults());
    }

    private ChannelComprehensionSummary saveResults(Consumer<DatabasePort> resultsConsumer) {
        TestDynamoDatabaseAdaptorComponent.FakeDynamoDBMapper mapper = (TestDynamoDatabaseAdaptorComponent.FakeDynamoDBMapper) component.getFakeDynamoDBMapper();
        resultsConsumer.accept(component.getDatabase());
        assertThat(mapper.getSaved()).hasSize(1);
        return mapper.getSaved().get(0);
    }
}

File: common/src/test/java/dev/number6/db/DynamoDatabaseAdaptorIntegrationTest.kt
package dev.number6.db

import dev.number6.db.TestDynamoDatabaseAdaptorComponent.FakeDynamoDBMapper
import dev.number6.db.model.ChannelComprehensionSummary
import dev.number6.db.port.DatabasePort
import dev.number6.generate.CommonRDG
import org.assertj.core.api.Assertions
import org.junit.jupiter.api.Test
import uk.org.fyodor.generators.RDG
import java.time.LocalDate
import java.util.function.Consumer

class DynamoDatabaseAdaptorIntegrationTest {
    private val component = DaggerTestDynamoDatabaseAdaptorComponent.create()

    @Test
    fun createChannels() {
        val channelNames: Collection<String> = RDG.list(RDG.string()).next()
        val mapper = component.fakeDynamoDBMapper as FakeDynamoDBMapper
        component.database.createNewSummaryForChannels(channelNames, LocalDate.now())
        Assertions.assertThat(mapper.saved).hasSize(channelNames.size)
                .extracting("channelName")
                .containsExactlyInAnyOrder(*channelNames.toTypedArray())
    }

    @Test
    fun saveEntityResults() {
        val results = CommonRDG.presentableEntityResults().next()
        val summary = saveResults(Consumer { m: DatabasePort? -> m!!.save(results!!) })
        Assertions.assertThat(summary!!.entityTotals).isEqualTo(results!!.results)
    }

    @Test
    fun saveSentimentResults() {
        val results = CommonRDG.presentableSentimentResults().next()
        val summary = saveResults(Consumer { m: DatabasePort? -> m!!.save(results!!) })
        Assertions.assertThat(summary!!.sentimentTotals).isEqualTo(results!!.sentimentTotals)
        Assertions.assertThat(summary.sentimentScoreTotals).isEqualTo(results.sentimentScoreTotals)
    }

    @Test
    fun saveKeyPhrasesResults() {
        val results = CommonRDG.presentableKeyPhrasesResults().next()
        val summary = saveResults(Consumer { m: DatabasePort? -> m!!.save(results!!) })
        Assertions.assertThat(summary!!.keyPhrasesTotals).isEqualTo(results!!.results)
    }

    private fun saveResults(resultsConsumer: Consumer<DatabasePort?>): ChannelComprehensionSummary? {
        val mapper = component.fakeDynamoDBMapper as FakeDynamoDBMapper
        resultsConsumer.accept(component.database)
        Assertions.assertThat(mapper.saved).hasSize(1)
        return mapper.saved[0]
    }
}